### ISSUE TITLE
The constructor to TemporayCredentials class can never return

### ISSUE THREAD

**Issue by addisonj** [aws/aws-sdk-js#1315]
This is partially related to #1064. 

When you create a `TemporaryCredentials` object after having resolved the default providers in `CredentialProviderChain` that ends up using an `EC2MetadataCredentials`, the constructor to `TemporaryCredentials` never returns.

Here a minimal test case that MUST be run on a real ec2 instance with no env credentials (must use `EC2MetadataCredentials`):

```JavaScript
const AWS = require('aws-sdk')

new AWS.CredentialProviderChain().resolve((err, creds) => {
  if (err) throw err
  console.log('have creds ' + !!creds)
  console.log('creating temp creds')
  const t = new AWS.TemporaryCredentials()
  console.log('never going to get here', t)
})

// need something to keep the event loop online
setInterval(() => {}, 100)
```

This was tested with latest sdk 2.7.25.

The reason I came across this:

I was hoping to avoid having to update the global credentials (`AWS.config.update({credentials: creds})` and instead simply set `temporaryCredentials.masterCredentials` directly (the API docs say its writable). However, due to this strange behavior, that isn't possible.

**Comment by chrisradek** [comment]
@addisonj 
`new AWS.TemporaryCredentials()` is actually throwing an error that's getting swallowed.
It currently requires for the global `AWS.config.credentials` to be set. 
The issue happens when trying to access `this.masterCredentials.masterCredentials` before `AWS.config.credentials` is set (and thus null):
https://github.com/aws/aws-sdk-js/blob/master/lib/credentials/temporary_credentials.js#L100

The reason this only happens with `EC2MetadataCredentials` is because the synchronous credential providers are resolved when the SDK is imported, and the global config is updated if credentials are found. Using `resolve`, it's a manual process.

The only work-around right now is to modify the global config. Given the unexpected behavior you're seeing, I'm going to take a look at your related issue again and see how much time it would take to implement.

**Comment by chrisradek** [comment]
@addisonj 
I created PR #1322 to allow passing master credentials to the TemporaryCredentials provider.

**Comment by addisonj** [comment]
This looks good to me!

**Comment by chrisradek** [comment]
Closing since the PR was released with version 2.7.28 of the SDK. Let us know if that works for you!
